crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -1134,6 +1084,82 @@ where
+    /// which returns true if the missing parent is sufficiently ahead of the canonical tip. If so,
+    /// the pipeline is run. Otherwise, we need to insert blocks using the blockchain tree, and
+    /// must download blocks outside of the pipeline. In this case, the distance is used to
+    /// determine how many blocks we should download at once.
+    fn on_disconnected_block(
+        &mut self,
+        downloaded_block: BlockNumHash,
+        missing_parent: BlockNumHash,
+    ) {
+        // compare the missing parent with the canonical tip
+        let canonical_tip_num = self.blockchain.canonical_tip().number;
+        let sync_target_state = self.forkchoice_state_tracker.sync_target_state();
+
+        trace!(target: "consensus::engine", ?downloaded_block, ?missing_parent, tip=?canonical_tip_num, "Handling disconnected block");
+
+        let mut exceeds_pipeline_run_threshold =
+            self.exceeds_pipeline_run_threshold(canonical_tip_num, missing_parent.number);
+
+        // check if the downloaded block is the tracked safe block
+        if let Some(ref state) = sync_target_state {
+            if downloaded_block.hash == state.finalized_block_hash {
+                // we downloaded the finalized block
+                exceeds_pipeline_run_threshold =
+                    self.exceeds_pipeline_run_threshold(canonical_tip_num, downloaded_block.number);
+            } else if let Some(buffered_finalized) =
+                self.blockchain.buffered_header_by_hash(state.finalized_block_hash)
+            {
+                // if we have buffered the finalized block, we should check how far
+                // we're off
+                exceeds_pipeline_run_threshold = self
+                    .exceeds_pipeline_run_threshold(canonical_tip_num, buffered_finalized.number);
+            }
+        }
+
+        // if the number of missing blocks is greater than the max, run the
+        // pipeline
+        if exceeds_pipeline_run_threshold {
+            if let Some(state) = sync_target_state {
+                // if we have already canonicalized the finalized block, we should
+                // skip the pipeline run
+                match self.blockchain.header_by_hash_or_number(state.finalized_block_hash.into()) {
+                    Err(err) => {
+                        warn!(target: "consensus::engine", ?err, "Failed to get safe block header");
+                    }
+                    Ok(None) => {
+                        // we don't have the safe block yet and the distance exceeds the allowed
+                        // threshold
+                        self.sync.set_pipeline_sync_target(state.safe_block_hash);
+                        // we can exit early here because the pipeline will take care of this
+                        return
+                    }
+                    Ok(Some(_)) => {
+                        // we're fully synced to the safe block
+                    }
```

Comment:
is this even reachable? under which conditions? should we still attempt to download the missing parent in this case?

Reply:
yes we should still attempt to download, because if we set the threshold to 32 but only check the finalized tag here, then this branch is common